packages/async-flow/src/replay-membrane.js

```diff
@@ -0,0 +1,393 @@
+      } else {
+        const entry = guestToHost(guestEntry);
+        log.pushEntry(entry);
+        const [_, ...args] = entry;
+        nestInterpreter(callIndex);
+        outcome = performCall(...args);
+      }
+    } catch (fatalError) {
+      throw panic(fatalError);
+    }
+
+    if (outcome.kind === 'return') {
+      return outcome.result;
+    } else {
+      outcome.kind === 'throw' ||
+        // @ts-expect-error TS correctly knows this case would be outside
+        // the type. But that's what we want to check.
+        Fail`unexpected outcome kind ${q(outcome.kind)}`;
+      throw outcome.problem;
+    }
+  };
+
+  // //////////////// Converters ///////////////////////////////////////////////
+
+  const makeGuestForHostRemotable = hRem => {
+    let gRem;
+    /** @param {PropertyKey} [optVerb] */
+    const makeGuestMethod = (optVerb = undefined) => {
+      const guestMethod = (...guestArgs) => {
+        const callIndex = log.getIndex();
+        return guestCallsHost(gRem, optVerb, guestArgs, callIndex);
+      };
+      if (optVerb) {
+        defineProperties(guestMethod, {
+          name: { value: String(optVerb) },
+          length: { value: Number(hRem[optVerb].length || 0) },
+        });
+      } else {
+        defineProperties(guestMethod, {
+          name: { value: String(hRem.name || 'anon') },
+          length: { value: Number(hRem.length || 0) },
+        });
+      }
+      return guestMethod;
+    };
+    const iface = String(getInterfaceOf(hRem) || 'remotable');
+    const guestIface = `${iface} guest wrapper`; // just for debugging clarity
+    if (typeof hRem === 'function') {
+      // NOTE: Assumes that a far function has no "static" methods. This
+      // is the current marshal design, but revisit this if we change our
+      // minds.
+      gRem = Remotable(guestIface, undefined, makeGuestMethod());
+      // NOTE: If we ever do support that, probably all we need
+      // to do is remove the following `throw Fail` line.
+      throw Fail`host far functions not yet passable`;
+    } else {
+      const methodNames = getMethodNames(hRem);
+      const guestMethods = methodNames.map(name => [
+        name,
+        makeGuestMethod(name),
+      ]);
+      // TODO in order to support E *well*,
+      // use HandledPromise to make gRem a remote presence for hRem
+      gRem = Remotable(guestIface, undefined, fromEntries(guestMethods));
+    }
+    return gRem;
+  };
+  harden(makeGuestForHostRemotable);
+
+  const makeGuestForHostVow = hVow => {
+    // TODO in order to support E *well*,
+    // use HandledPromise to make `promise` a handled promise for hVow
+    const { promise, resolve, reject } = makePromiseKit();
+    guestPromiseMap.set(promise, harden({ resolve, reject }));
+
+    watchWake(hVow);
+
+    void when(
+      hVow,
+      async hostFulfillment => {
+        await log.promiseReplayDone();
+        if (guestPromiseMap.get(promise) !== 'settled') {
```

Comment:
- need to check bijection for this promise
- need to clear guestPromiseMap when we clear ephemeral state
- should avoid redundant watch, but might require more bookkeeping
- analyze / explain fulfilled vow noticed but not recorded because it is during a failed replay
-

Reply:
Resolved base on verbal discussion w @mhofman

Reply:
Bug is fixed but still need explanation.

Reply:
The fix I believe hasn't been pushed yet and it requires to check `!stopped` here (and in the rejection handler)

Reply:
Here is the explanation I came up with. Please adapt as you see fit:

> The replay membrane is the only component inserting entries into the log. In particular, the flow's vow durable watcher does not log the settlement outcome, and instead it's the responsibility of the membrane's ephemeral handler. Because of this, the membrane's handler must be careful to:
> - Be added to the vow if the settlement has not yet been recorded in the log
> - Insert a single settlement outcome in the log for the given vow
> 
> In practice the former is accomplished by a handler always being added to the host vow when creating a guest promise, and the handler checking after replay is complete whether the guest promise is already settled (by the log replay) or not. The latter is accomplished by checking that the membrane has not been stopped before updating the log.

Reply:
Just included it literally. Done.